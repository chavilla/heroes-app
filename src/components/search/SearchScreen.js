import React, { useMemo } from "react";
import queryString from "query-string";
import { useLocation } from "react-router-dom";
import HeroCard from "../heroes/HeroCard";
import useForm from "../hooks/useForm";
import { getHeroByName } from "../selectors/getHeroByName";

const SearchScreen = ({ history }) => {
  // location hook
  const location = useLocation();

  //destructuring to the query
  const { q = "" } = queryString.parse(location.search);

  const [{ name }, handleChange] = useForm({ name: q });

  const heroesFiltered = useMemo(() => getHeroByName(q), [q]);

  const handleSubmit = (e) => {
    e.preventDefault();
    history.push(`?q=${name}`);
  };

  return (
    <div className="row">
      <div className="col-md-5">
        <h4>Search form</h4>
        <hr />
        <form onSubmit={handleSubmit}>
          <input
            autoComplete="off"
            type="text"
            placeholder="find your hero"
            className="form-control"
            name="name"
            onChange={handleChange}
            value={name}
          />
          <button className="btn mt-2 btn-outline-primary" type="submit">
            Search...
          </button>
        </form>
      </div>
      <div className="col-md-7">
        <h4>Results</h4>
        <hr />

        {q === "" && (
          <div className="alert-primary py-2 text-center">Search a hero</div>
        )}

        {q !== "" && heroesFiltered.length === 0 && (
          <div className="alert-danger py-2 text-center">
            There is no results with {q}
          </div>
        )}

        {heroesFiltered.map((hero) => (
          <HeroCard key={hero.id} {...hero} />
        ))}
      </div>
    </div>
  );
};

export default SearchScreen;
