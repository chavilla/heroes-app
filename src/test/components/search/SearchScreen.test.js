import React, { Component } from 'react';
import {mount} from 'enzyme';
import '@testing-library/jest-dom';
import {MemoryRouter, Route} from 'react-router-dom'
import SearchScreen from '../../../components/search/SearchScreen';

describe('<SearchScreen />', () => {

    const history={
        push:jest.fn()
    }

    const wrapper=mount(
        <MemoryRouter initialEntries={['/search']}>
            <Route 
            path='/search'
            component={()=> <SearchScreen history={history} />} 
            />
        </MemoryRouter>
    );

    test('should to display the component', () => {
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-primary').text()).toBe('Search a hero');
    });

    test('input should have batman', () => {
        const wrapper=mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                path='/search'
                component={()=> <SearchScreen history={history} />} 
                />
            </MemoryRouter>
        );
    
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });


    test('should have an error', () => {
        const wrapper=mount(
            <MemoryRouter initialEntries={['/search?q=batmanxhdwe']}>
                <Route 
                path='/search'
                component={()=> <SearchScreen history={history} />} 
                />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').exists()).toBe(true);
    });

    test('should have been called the history.push', () => {

        const wrapper=mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                path='/search'
                component={()=> 
                <SearchScreen history={history} />} 
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', { preventDefault: jest.fn()});

        wrapper.find('form').prop('onSubmit')({preventDefault: jest.fn() });
        expect(history.push).toHaveBeenCalledWith('?q=batman');
    })
    
    
    
})
    
    