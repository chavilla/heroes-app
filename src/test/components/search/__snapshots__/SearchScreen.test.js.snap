// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> should have an error 1`] = `
<div
  className="row"
>
  <div
    className="col-md-5"
  >
    <h4>
      Search form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="name"
        onChange={[Function]}
        placeholder="find your hero"
        type="text"
        value="batmanxhdwe"
      />
      <button
        className="btn mt-2 btn-outline-primary"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-md-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert-danger py-2 text-center"
    >
      There is no results with 
      batmanxhdwe
    </div>
  </div>
</div>
`;

exports[`<SearchScreen /> should to display the component 1`] = `
<div
  className="row"
>
  <div
    className="col-md-5"
  >
    <h4>
      Search form
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="name"
        onChange={[Function]}
        placeholder="find your hero"
        type="text"
        value=""
      />
      <button
        className="btn mt-2 btn-outline-primary"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-md-7"
  >
    <h4>
      Results
    </h4>
    <hr />
    <div
      className="alert-primary py-2 text-center"
    >
      Search a hero
    </div>
  </div>
</div>
`;
